File: dyros_robotics_suite/dyros_robot_controller/src/fr3_controller/robot_data.cpp
#include "fr3_controller/robot_data.h"


namespace FR3Controller
{
    RobotData::RobotData(const std::string& urdf_path, const bool verbose)
    {
        std::ifstream file(urdf_path);
        if (!file.good()) std::cout << "\033[1;31m" << "URDF file does not exist! : " << urdf_path << "\033[0m" << "\033[0m" << std::endl;
    
        pinocchio::urdf::buildModel(urdf_path, model_);
        data_ = pinocchio::Data(model_);
                
        joint_names_ = model_.names;
        joint_names_.erase(joint_names_.begin());  // Remove the first element "universe_joint"

        if(verbose)
        {
            std::cout << "Total nq = " << model_.nq << '\n' << "Total nv = " << model_.nv << "\n\n";
            std::cout << " id | name              | nq | nv | idx_q | idx_v\n";
            std::cout << "----+-------------------+----+----+-------+------\n";
            for(pinocchio::JointIndex id = 1; id < model_.joints.size(); ++id)
            {
                std::cout << std::setw(3)  << id << " | "
                          << std::setw(17) << model_.names[id] << " | "
                          << std::setw(2)  << model_.nqs[id]   << " | "
                          << std::setw(2)  << model_.nvs[id]   << " | "
                          << std::setw(5)  << model_.idx_qs[id]<< " | "
                          << std::setw(4)  << model_.idx_vs[id]<< '\n';
            }
        }


        // Initialize joint space state
        q_.setZero(model_.nq);
        qdot_.setZero(model_.nv);

        // Initialize task space state
        x_ .setIdentity();
        xdot_.setZero();
        J_.setZero(6, model_.nv);
        Jdot_.setZero(6, model_.nv);

        // Initialize joint space dynamics
        M_.setZero(model_.nv, model_.nv);
        M_inv_.setZero(model_.nv, model_.nv);
        g_.setZero(model_.nv);       
        c_.setZero(model_.nv);       
        NLE_.setZero(model_.nv);       
    }

    RobotData::~RobotData()
    {
    }

    bool RobotData::updateState(const VectorXd& q, const VectorXd& qdot)
    {
        q_ = q;
        qdot_ = qdot;
        if(!updateKinematics(q_, qdot_)) return false;
        if(!updateDynamics(q_, qdot_)) return false;
        return true;
    }

    bool RobotData::updateKinematics(const VectorXd& q, const VectorXd& qdot)
    {
        if(q.size() != model_.nq)
        {
            std::cerr << "\033[1;31m" << "updateDynamics Error: size of q " << q.size() << " is not equal to model.nq size: " << model_.nq << "\033[0m" << std::endl;
            return false;
        }
        if(qdot.size() != model_.nv)
        {
            std::cerr << "\033[1;31m" << "updateDynamics Error: size of qdot " << qdot.size() << " is not equal to model.nv size: " << model_.nv << "\033[0m" << std::endl;
            return false;
        }
        
        pinocchio::FrameIndex link_index = model_.getFrameId(ee_name_);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << ee_name_ << " not found in URDF." << "\033[0m" << std::endl;
            return false;
        }

        pinocchio::computeJointJacobians(model_, data_, q);
        pinocchio::computeJointJacobiansTimeVariation(model_, data_, q, qdot);
        x_ = getPose(ee_name_);
        J_ = getJacobian(ee_name_);
        xdot_ = J_ * qdot;
        Jdot_ = getJacobianTimeVariation();

        return true;
    }

    bool RobotData::updateDynamics(const VectorXd& q, const VectorXd& qdot)
    {
        if(q.size() != model_.nq)
        {
            std::cerr << "\033[1;31m" << "updateDynamics Error: size of q " << q.size() << " is not equal to model.nq size: " << model_.nq << "\033[0m" << std::endl;
            return false;
        }
        if(qdot.size() != model_.nv)
        {
            std::cerr << "\033[1;31m" << "updateDynamics Error: size of qdot " << qdot.size() << " is not equal to model.nv size: " << model_.nv << "\033[0m" << std::endl;
            return false;
        }
        pinocchio::crba(model_, data_, q);
        pinocchio::computeGeneralizedGravity(model_, data_, q);
        pinocchio::nonLinearEffects(model_, data_, q, qdot);

        // update joint space dynamics
        M_ = data_.M;
        M_ = M_.selfadjointView<Upper>();  // Only upper triangular part of M_ is computed by pinocchio::crba
        M_inv_ = M_.inverse();
        g_ = data_.g;
        NLE_ = data_.nle;
        c_ = NLE_ - g_;

        return true;
    }

    Matrix4d RobotData::computePose(const VectorXd& q, const std::string& link_name)
    {
        assert(q.size() == model_.nq);

        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return Matrix4d::Identity();
        }
        pinocchio::Data tmp_data;
        pinocchio::framesForwardKinematics(model_, tmp_data, q);
        Matrix4d link_pose = tmp_data.oMf[link_index].toHomogeneousMatrix();

        return link_pose;
    }

    Matrix4d RobotData::getPose(const std::string& link_name)
    {
        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return Matrix4d::Identity();
        }
        Matrix4d link_pose = data_.oMf[link_index].toHomogeneousMatrix();

        return link_pose;
    }

    MatrixXd RobotData::computeJacobian(const VectorXd& q, const std::string& link_name)
    {
        assert(q.size() == model_.nq);

        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return MatrixXd::Zero(6, model_.nv);
        }
        MatrixXd J;
        J.setZero(6, model_.nv);
        pinocchio::Data tmp_data;
        pinocchio::computeFrameJacobian(model_, tmp_data, q, link_index, J);

        return J;
    }

    MatrixXd RobotData::getJacobian(const std::string& link_name)
    {
        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return MatrixXd::Zero(6, model_.nv);
        }
        MatrixXd J = pinocchio::getFrameJacobian(model_, data_, link_index, pinocchio::ReferenceFrame::WORLD);

        return J;
    }

    MatrixXd RobotData::computeJacobianTimeVariation(const VectorXd& q, const VectorXd& qdot, const std::string& link_name)
    {
        assert(q.size() == model_.nq);
        assert(qdot.size() == model_.nv);

        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return MatrixXd::Zero(6, model_.nv);
        }
        MatrixXd Jdot;
        Jdot.setZero(6, model_.nv);
        pinocchio::Data tmp_data;
        pinocchio::computeJointJacobiansTimeVariation(model_, tmp_data, q, qdot_);
        pinocchio::getFrameJacobianTimeVariation(model_, tmp_data, link_index, pinocchio::ReferenceFrame::WORLD, Jdot);

        return Jdot;
    }

    MatrixXd RobotData::getJacobianTimeVariation(const std::string& link_name)
    {
        pinocchio::FrameIndex link_index = model_.getFrameId(link_name);
        if (link_index == static_cast<pinocchio::FrameIndex>(-1))  
        {
            std::cerr << "\033[1;31m" << "Error: Link name " << link_name << " not found in URDF." << "\033[0m" << std::endl;
            return MatrixXd::Zero(6, model_.nv);
        }
        MatrixXd Jdot;
        Jdot.setZero(6, model_.nv);
        pinocchio::getFrameJacobianTimeVariation(model_, data_, link_index, pinocchio::ReferenceFrame::WORLD, Jdot);

        return Jdot;
    }

    Vector6d RobotData::computeVelocity(const VectorXd& q, const VectorXd& qdot, const std::string& link_name)
    {
        assert(q.size() == model_.nq);
        assert(qdot.size() == model_.nv);

        MatrixXd J = computeJacobian(q, link_name);
        
        return J * qdot;
    }

    Vector6d RobotData::getVelocity(const std::string& link_name)
    {
        MatrixXd J = getJacobian(link_name);

        return J * qdot_;
    }

    MatrixXd RobotData::computeMassMatrix(const VectorXd& q)
    {
        assert(q.size() == model_.nq);
        
        pinocchio::Data tmp_data;
        pinocchio::crba(model_, tmp_data, q);
        tmp_data.M = tmp_data.M.selfadjointView<Upper>();  // Only upper triangular part of M_ is computed by pinocchio::crba

        return tmp_data.M;
    }

    VectorXd RobotData::computeCoriolis(const VectorXd& q, const VectorXd& qdot)
    {
        assert(q.size() == model_.nq);
        assert(qdot.size() == model_.nv);
        
        pinocchio::Data tmp_data;
        pinocchio::computeCoriolisMatrix(model_, tmp_data, q, qdot);

        return tmp_data.C * qdot;
    }

    VectorXd RobotData::computeGravity(const VectorXd& q)
    {
        assert(q.size() == model_.nq);
        
        pinocchio::Data tmp_data;
        pinocchio::computeGeneralizedGravity(model_, tmp_data, q);

        return tmp_data.g;
    }

    VectorXd RobotData::computeNonlinearEffects(const VectorXd& q, const VectorXd& qdot)
    {
        assert(q.size() == model_.nq);
        assert(qdot.size() == model_.nv);

        pinocchio::Data tmp_data;
        pinocchio::nonLinearEffects(model_, tmp_data, q, qdot);

        return tmp_data.nle;
    }
}

File: dyros_robotics_suite/dyros_robot_controller/dyros_robot_controller/ur5e_controller/robot_data.py
import numpy as np
import os
from rclpy.node import Node
from ament_index_python.packages import get_package_share_directory
import pinocchio as pin

"""
UR5e URDF Joint Information
Total nq = 6, total nv = 6
 id | name                 | nq | nv | idx_q | idx_v
----+----------------------+----+----+-------+------
  1 | shoulder_pan_joint   |  1 |  1 |     0 |    0
  2 | shoulder_lift_joint  |  1 |  1 |     1 |    1
  3 | elbow_joint          |  1 |  1 |     2 |    2
  4 | wrist_1_joint        |  1 |  1 |     3 |    3
  5 | wrist_2_joint        |  1 |  1 |     4 |    4
  6 | wrist_3_joint        |  1 |  1 |     5 |    5

"""

class UR5eRobotData():
    def __init__(self, node: Node, mj_joint_names: list):
        self.node = node
        self.mj_joint_names = mj_joint_names
        
        urdf_path = os.path.join(
            get_package_share_directory('dyros_robot_controller'),
            'robot',
            'ur5e.urdf'
        )
        # Build model and data
        self.model = pin.buildModelFromUrdf(urdf_path)
        self.data = pin.Data(self.model)
        self.ee_name = "flange"  # End-effector name in the URDF

        # Joint names (exclude universe joint at index 0)
        self.rd_joint_names = [f.name for f in self.model.frames if f.type == pin.JOINT]
        self.nq = self.model.nq
        self.nv = self.model.nv

        self.node.get_logger().info(f"Total nq = {self.nq}, total nv = {self.nv}")
        self.node.get_logger().info(" id | name                 | nq | nv | idx_q | idx_v")
        self.node.get_logger().info("----+----------------------+----+----+-------+------")
        for idx in range(1, len(self.model.joints)):
            joint = self.model.joints[idx]
            name = self.model.names[idx]
            self.node.get_logger().info(f"{idx:3d} | {name:20s} | {self.model.nqs[idx]:2d} | {self.model.nvs[idx]:2d} |")

        # Initialize states
        self.q = np.zeros(self.nq)
        self.qdot = np.zeros(self.nv)
        # Initialize kinematics
        self.x = np.eye(4)
        self.xdot = np.zeros(6)
        self.J = np.zeros((6, self.nv))
        self.Jdot = np.zeros((6, self.nv))
    
    def updateState(self, 
                    pos_dict: dict, 
                    vel_dict: dict, 
                    tau_ext_dict: dict, 
                    sensor_dict:dict):
        for rd_joint_name in self.rd_joint_names:
            rd_index = self.rd_joint_names.index(rd_joint_name)
            
            self.q[rd_index] = pos_dict[rd_joint_name]
            self.qdot[rd_index] = vel_dict[rd_joint_name]
        
        if(not self.updateKinematics(self.q, self.qdot)):
            self.node.get_logger().error("[UR5eRobotData] Failed to update robot state.")

# ==================================================================================================
    def updateKinematics(self, q: np.ndarray, qdot: np.ndarray) -> bool:
        """
        Update kinematic state: pose, Jacobian, and their time derivatives.

        Args:
            q (np.ndarray): Joint positions, shape (nq,).
            qdot (np.ndarray): Joint velocities, shape (nv,).

        Returns:
            bool: False if input sizes mismatch or frame not found, True otherwise.
        """
        if q.size != self.nq or qdot.size != self.nv:
            self.node.get_logger().error(f"[UR5eRobotData] Error: q size {q.size}, qdot size {qdot.size} do not match model dimensions.")
            return False

        # Update internal state
        self.q = q.copy()
        self.qdot = qdot.copy()

        # Compute joint Jacobians and their time variation
        pin.computeJointJacobians(self.model, self.data, self.q)
        pin.computeJointJacobiansTimeVariation(self.model, self.data, self.q, self.qdot)

        # Update frame placements for pose
        pin.framesForwardKinematics(self.model, self.data, self.q)

        # Get end-effector frame index
        frame_id = self.model.getFrameId(self.ee_name)
        if frame_id < 0:
            print(f"Error: Frame '{self.ee_name}' not found in model.")
            return False

        # Pose: homogeneous matrix
        self.x = self.data.oMf[frame_id].homogeneous
        # Jacobian in world frame
        self.J = pin.getFrameJacobian(self.model, self.data, frame_id, pin.ReferenceFrame.WORLD)
        # Spatial velocity
        self.xdot = self.J.dot(self.qdot)
        # Jacobian time derivative
        self.Jdot = pin.getFrameJacobianTimeVariation(self.model, self.data, frame_id, pin.ReferenceFrame.WORLD)

        return True
    
    def computePose(self, q: np.ndarray) -> np.ndarray:
        """
        Compute end-effector pose for given configuration.

        Args:
            q (np.ndarray): Joint positions, shape (nq,).

        Returns:
            np.ndarray: 4x4 homogeneous transform.
        """
        data_tmp = pin.Data(self.model)
        pin.framesForwardKinematics(self.model, data_tmp, q)
        frame_id = self.model.getFrameId(self.ee_name)
        if frame_id < 0:
            raise ValueError(f"Frame '{self.ee_name}' not found.")
        return data_tmp.oMf[frame_id].homogeneous

    def computeJacobian(self, q: np.ndarray) -> np.ndarray:
        """
        Compute end-effector Jacobian for given configuration.
        """
        data_tmp = pin.Data(self.model)
        pin.computeJointJacobians(self.model, data_tmp, q)
        frame_id = self.model.getFrameId(self.ee_name)
        if frame_id < 0:
            raise ValueError(f"Frame '{self.ee_name}' not found.")
        return pin.getFrameJacobian(self.model, data_tmp, frame_id, pin.ReferenceFrame.WORLD)

    def computeJacobianTimeVariation(
        self, q: np.ndarray, qdot: np.ndarray
    ) -> np.ndarray:
        """
        Compute time derivative of Jacobian for given state.
        """
        data_tmp = pin.Data(self.model)
        pin.computeJointJacobians(self.model, data_tmp, q)
        pin.computeJointJacobiansTimeVariation(self.model, data_tmp, q, qdot)
        frame_id = self.model.getFrameId(self.ee_name)
        if frame_id < 0:
            raise ValueError(f"Frame '{self.ee_name}' not found.")
        return pin.getFrameJacobianTimeVariation(
            self.model, data_tmp, frame_id, pin.ReferenceFrame.WORLD
        )

    def computeVelocity(self, q: np.ndarray, qdot: np.ndarray) -> np.ndarray:
        """
        Compute end-effector spatial velocity J(q)*qdot.
        """
        J = self.computeJacobian(q)
        return J.dot(qdot)

    def getPose(self) -> np.ndarray:
        """
        Get the current end-effector pose.

        Returns:
            np.ndarray: 4x4 homogeneous transformation matrix.
        """
        return self.x.copy()

    def getJacobian(self) -> np.ndarray:
        """
        Get the current end-effector Jacobian.

        Returns:
            np.ndarray: 6xnv Jacobian matrix.
        """
        return self.J.copy()

    def getJacobianTimeVariation(self) -> np.ndarray:
        """
        Get the time derivative of the end-effector Jacobian.

        Returns:
            np.ndarray: 6xnv Jacobian time derivative.
        """
        return self.Jdot.copy()

    def getVelocity(self) -> np.ndarray:
        """
        Get the current end-effector spatial velocity.

        Returns:
            np.ndarray: 6D spatial velocity vector.
        """
        return self.xdot.copy()
